✨ Implement auto-recovery of the managed TGIS subprocess

tjohnson31415 opened the pull request:
Before this change, the TGIS server was launched in a sub-process and the code assumed that it never crashed. Now, the code detects when requests are failing and will restart the TGIS subprocess. The restart happens in a background thread so as to not block any request threads, which means that requests will still fail as the process is restarted

High level changes:
- refactor the managed TGIS subprocess to a new module
- if inference requests to TGIS return failures, a health check is sent to the subprocess and it is restarted if the health check fails
- the transition from BOOTING to READY for the subprocess is now tracked by a background daemon thread
    - this is to de-couple the request threads from the server pool from the booting of the subprocess
- wrap the gRPC client returned by `get_client(model)` into a class so that the underlying gRPC client can be re-created
    - a typical gRPC client is created with an existing channel and will attempt to reconnect with a back-off, but I think it would be better to manage the connection explicitly since we also manage the server that it connects to.
    - the wrapping layer is also where errors TGIS are checked to determine when to restart the process
- update test mocks to enable testing the auto-recovery
    - fully expose the mock servers in TGISMock (not just connection information) so that the test can stop the servers
    - combine the components of the subprocess mocking into MockTGISFixture
    - have the TGISMock server(s) boot triggered by the code as a "side effect" of the `subprocess.Popen` mock
    - allow the mock servers to be stop'd and start'd repeatedly (creating a new thread on start)

tjohnson31415 committed:
refactor: extract managed tgis to separate module

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test/refactor: rework tgis_mock to expose servers

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: make TGISMock restartable

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: refactor tgis backend tests

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: add autorecovery test

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
feat: implement autorecovery for the TGIS subprocess

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
🎨 cleanup, format, and lint

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

gkumbhat reviewed (commented):
meta comment: while this is in progress, I think it would be great to have a diagram in readme somewhere that describes the functioning of the backend, since it involves both threading and sub-processing working together although for different reasons. Having a diagram would help future readers.

tjohnson31415 committed:
remove use of typing.Self that is 3.11 only

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+
+
+# pylint: disable=too-many-instance-attributes
+class ManagedTGISSubprocess:
```
NIT: It would be nice to have a docstring. If this should be the same as the module-level docstring, you can cheat by doing the following:

```py
class ManagedTGISSubproces:
    __doc__ = __doc__
    ...
```

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+        health_poll_timeout: Optional[float] = None,
+        bootup_poll_delay: Optional[float] = None,
+        load_timeout: Optional[float] = None,
+    ):
```
NIT: docstring here too would be good for things like units on arguments

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+        bootup_poll_delay: Optional[float] = None,
+        load_timeout: Optional[float] = None,
+    ):
+        if bootup_poll_delay is None:
```
Is this done so that people can _explicitly_ set `None` to get the default? Curious why this isn't just done as defaults in the args since they're all primitives.

tjohnson31415 replied:
Ah, yes, I'll remove these checks.
I had these parameters as `datetime.timedelta` instances for a while and I wasn't sure about the mutability of them, but I forgot to clean this up when I removed the `timedelta`s.

gabe-l-hart replied:
ah, makes total sense

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+        self._ensure_terminated()
+
+        # Launch TGIS
+        launch_cmd = " ".join(
```
NIT: I'm wondering if any of this can be stored on `self` to avoid some of the extra instance variables (pesky `pylint`!)

tjohnson31415 replied:
Hmm, of course some of it could be. When I run `python -m tox -e lint` I don't see these lines being picked out. What's the lint rule you are seeing? Is there a change that should be made to `.pylintrc` to check for it?

gabe-l-hart replied:
I was looking at

```py
# pylint: disable=too-many-instance-attributes
class ManagedTGISSubprocess:
```

tjohnson31415 replied:
Ah, well, with that, pylint is happy ;)

I may be misunderstanding your first comment, but that linting rule is about having [too many attributes on self](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-instance-attributes.html). Storing something like `launch_cmd` on `self` is the opposite of what it wants.

EDIT: NVM, I think I understand your meaning now. `launch_cmd` takes two instance vars, so we could reduce the count by 1 by having `self._launch_cmd` instead of `self._model_path` and `self._http_port`

---

I'll see if I can come up with a nice way to combine some of the instance attributes. I could throw in some dataclasses for the related ones.

gabe-l-hart replied:
Heh, yeah, sorry that was just poorly written. You're right, I was wondering if we could reduce the count by combining them into pre-built formatted strings for the commands. _very_ low priority!

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+        self._tgis_state = _TGISState.BOOTING
+        # launch background thread to monitor the boot sequence
+        # save a handle to it to record bootup errors
+        self._bootup_thread = threading.Thread(target=self._monitor_bootup, daemon=True)
```
Interesting! I've never actually used `daemon` like this, but reading [the docs](https://docs.python.org/3/library/threading.html#thread-objects) this makes perfect sense

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+                        )
+                        self._bootup_exc = exc
+                        error("<MTS11752287E>", exc)
+                except AttributeError:
```
Is there a specific `AttributeError` that you plan to catch here? It's not immediately clear to me why this would happen

tjohnson31415 replied:
The `AttributeError` would arise if `self._tgis_proc` is `None` in the `if self._tgis_proc.poll()` conditional, which could happen if the subprocess was terminated with the `_monitor_bootup()` thread running in the background.

I'll make that check more explicit though and remove the try-catch.

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+            if time.time() - start_t >= self._load_timeout:
+                with self._mutex:
+                    # double check that we are still in the booting state
+                    if not self._tgis_state == _TGISState.BOOTING:
```
Curious why `not X == Y` vs `X != Y`? No real preference on my end

tjohnson31415 replied:
Writing code is sometimes a stochastic process with environmental factors, aka I also didn't have a strong preference and didn't choose consistently :P

I'll make em `X != Y`.

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+                ).start()
+                raise rpc_error
+
+            exc_chain = RuntimeError("Unhandled RPC error")
```
I'm not sure I'm totally following the logic here. Why is `rpc_error` what gets raised instead of `exc_chain`? Also, I think the `error` call will itself `raise`, so the following `raise rpc_error` line will never be hit (unless errors are globally disabled)?

tjohnson31415 replied:
I'll take another pass at this section, there's probably some more cases I should be handling and comments to be added.

> Why is rpc_error what gets raised instead of exc_chain?

The idea behind adding using `exc_chain` is to add an extra bit of information to indicate that the wrapped client did not have a branch for the particular kind of error, which also means that no auto-recovery is triggered. For the UNAVAILABLE case, the error is "handled" so I didn't wrap it up in the exception chain. Also, I just noticed that I can include a `root_exception` in the call to `error` instead of doing that manually like I did.

> Also, I think the error call will itself raise, so the following raise rpc_error line will never be hit ?

Ah, yep good point. That reminds me of a question I had that I didn't see from a glance at the `error_handler`:
Can I plug into the error handler's logging without raising? Should I call `_handle_exception_messages` directly?

> (unless errors are globally disabled)

There is a way to cause the `error_handler` to not raise? If so, wouldn't an explicit `return` after it be good to have where flow control is needed?

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+                # raise the bootup exception if we captured one
+                if self._bootup_exc:
+                    raise self._bootup_exc
+                return False
```
What is the `bool` return type used for here?

tjohnson31415 replied:
Hmm, yeah... its a bit awkward currently... I'll update this function to either return if ready or throw an exception if not instead of having some cases where it returns a boolean False.

gabe-l-hart commented on caikit_tgis_backend/managed_tgis_subprocess.py:
```diff
@@ -0,0 +1,371 @@
+            if self._tgis_proc:
+                self._tgis_proc.terminate()
+                self._tgis_proc = None
+                # disown the bootup thread (but keep the exception if one was captured)
```
NIT: maybe I'm just missing something, but I'm not seeing where the exception is captured?

tjohnson31415 replied:
`self._bootup_exc` is set in the `_monitor_bootup()` thread to record why the boot up failed.
The comment here is meant to explain why there is no `self._bootup_exc = None` line. I'll make that clearer:
```
# (but keep self._bootup_exc if one was captured)
```

gabe-l-hart commented on tests/test_tgis_backend.py:
```diff
@@ -220,44 +251,107 @@ def test_local_tgis_unload(mock_tgis_proc_ok, tgis_mock_insecure):
+
+    # wait for the server to reboot
+    # pause this thread to allow the reboot thread to start
+    time.sleep(0.5)
```
Curious why we need this sleep here?

tjohnson31415 replied:
The restart of TGIS is triggered by the failed request which launches a background thread. If `wait_until_ready()` is called before the background thread detects that TGIS is down, it will just return READY. There's probably a better way, but the `sleep` is meant to force a context switch so that the background thread gets a chance to detect the failure.

gabe-l-hart replied:
got it, makes sense!

gabe-l-hart commented on tests/tgis_mock.py:
```diff
@@ -128,110 +109,171 @@
+        if self.ca_cert:
+            self.ca_cert_file = os.path.join(self._tls_dir.name, "ca.pem")
+            with open(self.ca_cert_file, "w") as handle:
+                handle.write(self.ca_cert)
```
Fascinating! This does indeed save a cert to disk, but it's an auto-generated cert that is scoped to this test and useless elsewhere. I'm genuinely curious whether this is just keying off the name `ca_cert`?

gabe-l-hart reviewed (commented):
Wow, Travis, this is _much_ better than what was there before! Nice job making the changes transparent to the interface too. I've got a couple of little NITs, mostly around comments, docstrings, and error messages, but nothing is a showstopper.

tjohnson31415 committed:
review: cleanup from review comments

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
doc: add missing docstrings and improve comments

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
test: update tgis mock health check configs

The test_local_tgis_run test started failing for me with the short
timeout and frequent health requests...

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
review: small refactor of wait_until_ready

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
review: extra if instead of try-catch

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
review: simplify rpc error checking

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
lint: wrap docstrings for pylint

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

tjohnson31415 committed:
Merge branch 'main' into subprocess-autorecovery

* main:
  :art: fmt
  :bug: whoops, use hasattr
  :bug: remove internal dict ref for backends
  :arrow_up: bump caikit
  :arrow_up: Update caikit requirement

tjohnson31415 committed:
fix: add back num_gpus parameter

Signed-off-by: Travis Johnson <tsjohnso@us.ibm.com>

gabe-l-hart reviewed (approved):
Looks great!
